File: src/main/java/view/EditView.java
// EditView.java
//
// Copyright 2002-2022 Jack Boyce and the Juggling Lab contributors

package jugglinglab.view;

import java.awt.*;
import java.io.File;
import java.util.*;
import javax.swing.*;
import javax.swing.border.*;

import jugglinglab.core.*;
import jugglinglab.jml.*;
import jugglinglab.util.*;


// This view provides the ability to edit a pattern visually. It features a
// ladder diagram on the right and an animator on the left.

public class EditView extends View {
    protected AnimationEditPanel aep;
    protected JPanel ladder;
    protected JSplitPane jsp;


    public EditView(Dimension dim, JMLPattern pat) throws
                            JuggleExceptionUser, JuggleExceptionInternal {
        aep = new AnimationEditPanel();
        aep.setPreferredSize(dim);
        aep.setMinimumSize(new Dimension(10, 10));

        ladder = new JPanel();
        ladder.setLayout(new BorderLayout());
        ladder.setBackground(Color.white);

        // add ladder diagram now to get dimensions correct; will be replaced
        // in restartView()
        ladder.add(new LadderDiagram(pat), BorderLayout.CENTER);

        Locale loc = Locale.getDefault();
        if (ComponentOrientation.getOrientation(loc) == ComponentOrientation.LEFT_TO_RIGHT) {
            jsp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, aep, ladder);
            jsp.setResizeWeight(1.0);
        } else {
            jsp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ladder, aep);
            jsp.setResizeWeight(0.0);
        }
        jsp.setBorder(new EmptyBorder(0, 0, 0, 0));
        jsp.setBackground(Color.white);

        setBackground(Color.white);
        setLayout(new BorderLayout());
        add(jsp, BorderLayout.CENTER);
    }

    // View methods

    @Override
    public void restartView(JMLPattern p, AnimationPrefs c) throws
                            JuggleExceptionUser, JuggleExceptionInternal {
        boolean changing_jugglers = (p != null && getPattern() != null &&
                  p.getNumberOfJugglers() != getPattern().getNumberOfJugglers());

        aep.restartJuggle(p, c);
        setAnimationPanelPreferredSize(getAnimationPrefs().getSize());

        if (p != null) {
            EditLadderDiagram new_ladder = new EditLadderDiagram(p, parent, this);
            new_ladder.setAnimationPanel(aep);
            // LadderDiagram new_ladder = new LadderDiagram(p);
            // new_ladder.setAnimator(aep.getAnimator());

            aep.addAnimationAttachment(new_ladder);
            aep.deactivateEvent();
            aep.deactivatePosition();

            ladder.removeAll();
            ladder.add(new_ladder, BorderLayout.CENTER);

            if (changing_jugglers && parent != null) {
                // the next line is needed to get the JSplitPane divider to
                // reset during layout
                jsp.resetToPreferredSizes();

                if (parent.isWindowMaximized())
                    parent.validate();
                else
                    parent.pack();
            } else
                ladder.validate();  // to make ladder redraw

            if (parent != null)
                parent.setTitle(p.getTitle());
        }
    }

    @Override
    public void restartView() throws JuggleExceptionUser, JuggleExceptionInternal {
        aep.restartJuggle();
    }

    @Override
    public Dimension getAnimationPanelSize() {
        return aep.getSize(new Dimension());
    }

    @Override
    public void setAnimationPanelPreferredSize(Dimension d) {
        aep.setPreferredSize(d);
        jsp.resetToPreferredSizes();
    }

    @Override
    public JMLPattern getPattern() {
        return aep.getPattern();
    }

    @Override
    public AnimationPrefs getAnimationPrefs() {
        return aep.getAnimationPrefs();
    }

    @Override
    public double getZoomLevel() {
        return aep.getZoomLevel();
    }

    @Override
    public void setZoomLevel(double z) {
        aep.setZoomLevel(z);
    }

    @Override
    public boolean isPaused() {
        return aep.isPaused();
    }

    @Override
    public void setPaused(boolean pause) {
        if (aep.message == null)
            aep.setPaused(pause);
    }

    @Override
    public void disposeView() {
        aep.disposeAnimation();
    }

    @Override
    public void writeGIF(File f) {
        aep.writingGIF = true;
        boolean origpause = isPaused();
        setPaused(true);
        jsp.setEnabled(false);
        if (parent != null)
            parent.setResizable(false);

        Runnable cleanup = new Runnable() {
            @Override
            public void run() {
                aep.writingGIF = false;
                setPaused(origpause);
                jsp.setEnabled(true);
                if (parent != null)
                    parent.setResizable(true);
            }
        };

        new View.GIFWriter(aep, f, cleanup);
    }
}

File: composeApp/src/jvmMain/kotlin/jugglinglab/view/EditView.kt
//
// EditView.kt
//
// This view provides the ability to edit a pattern visually. It features a
// ladder diagram on the right and an animator on the left.
//
// Copyright 2002-2026 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.view

import jugglinglab.core.*
import jugglinglab.jml.JmlPattern
import jugglinglab.ui.AnimationPanel
import jugglinglab.ui.LadderDiagramPanel
import jugglinglab.ui.PatternWindow
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import java.awt.BorderLayout
import java.awt.Color
import java.awt.ComponentOrientation
import java.awt.Dimension
import java.util.Locale
import javax.swing.JSplitPane
import javax.swing.border.EmptyBorder

class EditView(
    state: PatternAnimationState,
    patternWindow: PatternWindow
) : View(state, patternWindow) {
    private val ap = AnimationPanel(state)
    private val ladder = LadderDiagramPanel(state, patternWindow)
    private val jsp: JSplitPane

    init {
        ap.preferredSize =  if (patternWindow.isWindowMaximized) {
            // leave enough room for preferred width of ladder; layout
            // will expand the animator dimensions to fit
            Dimension(patternWindow.width * 3 / 4, 50)
        } else {
            Dimension(state.prefs.width, state.prefs.height)
        }
        ap.minimumSize = Dimension(50, 50)

        val loc = Locale.getDefault()
        if (ComponentOrientation.getOrientation(loc) == ComponentOrientation.LEFT_TO_RIGHT) {
            jsp = JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ap, ladder)
            jsp.setResizeWeight(1.0)
        } else {
            jsp = JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ladder, ap)
            jsp.setResizeWeight(0.0)
        }
        jsp.setBorder(EmptyBorder(0, 0, 0, 0))
        jsp.setBackground(Color.white)

        setBackground(Color.white)
        setLayout(BorderLayout())
        add(jsp, BorderLayout.CENTER)
    }

    //--------------------------------------------------------------------------
    // View methods
    //--------------------------------------------------------------------------

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    override fun restartView(pattern: JmlPattern?, prefs: AnimationPrefs?, coldRestart: Boolean) {
        val sizeChanged = (prefs != null && (prefs.width != state.prefs.width || prefs.height != state.prefs.height))

        ap.restartJuggle(pattern, prefs, coldRestart)
        if (sizeChanged) {
            // The containing window will do a layout (validate() or pack()) in
            // PatternWindow.doMenuCommand(MenuCommand.VIEW_ANIMPREFS). Before
            // that, here we set the panels' preferred sizes so the layout
            // manager will allocate the right amount of space.
            setAnimationPanelPreferredSize(
                Dimension(state.prefs.width, state.prefs.height)
            )
            ladder.preferredSize = Dimension(ladder.size.width, state.prefs.height)

            // This makes the JSplitPane divider reset during layout
            jsp.resetToPreferredSizes()
        }
        if (pattern != null) {
            patternWindow.setTitle(pattern.title)
            patternWindow.updateColorsMenu()
        }
    }

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    override fun restartView() {
        ap.restartJuggle()
    }

    override val animationPanelSize: Dimension?
        get() = ap.getSize(Dimension())

    override fun setAnimationPanelPreferredSize(d: Dimension) {
        ap.preferredSize = d
    }

    override fun disposeView() {
        ap.disposeAnimation()
    }
}
